lib/fluent/plugin/filter_kubernetes_metadata.rb

```diff
@@ -224,40 +232,49 @@ def configure(conf)
 
     end
 
-    def get_metadata_for_record(namespace_name, pod_name, container_name)
+    def get_metadata_for_record(match_data)
+      namespace_name = match_data['namespace']
+      pod_name = match_data['pod_name']
       metadata = {
-        'container_name' => container_name,
+        'container_name' => match_data['container_name'],
         'namespace_name' => namespace_name,
-        'pod_name'       => pod_name,
+        'pod_name'       => pod_name
       }
       if @kubernetes_url.present?
-        cache_key = "#{namespace_name}_#{pod_name}"
-
-        this = self
-        pod_metadata = @cache.getset(cache_key) {
-          md = this.get_pod_metadata(
-            namespace_name,
-            pod_name,
-          )
-          md
-        }
+        cache_key = match_data[@cache_key]
+
+        pod_metadata = get_pod_metadata(cache_key, namespace_name, pod_name)
         metadata.merge!(pod_metadata) if pod_metadata
+      end
+      metadata
+    end
 
-        if @include_namespace_metadata
-          namespace_metadata = @namespace_cache.getset(namespace_name) {
-            begin
-              namespace = @client.get_namespace(namespace_name)
-              if namespace
-                parse_namespace_metadata(namespace)
-              end
-            rescue KubeException
-              nil
+    def get_pod_metadata(key, namespace_name, pod_name)
+        ids = @id_cache.getset(key) do 
+          @stats.bump(:pod_cache_miss)
+          pod_metadata = fetch_pod_metadata(namespace_name, pod_name)
+          namespace_metadata = fetch_namespace_metadata(namespace_name)
```

Comment:
If the `pod_metadata` is returned as an empty hash, do we want to always issue the fetch of the namespace name?